frontend/server/minio-helper.ts

```diff
@@ -54,24 +56,80 @@ export async function createMinioClient(config: MinioClientOptionsWithOptionalSe
   return new MinioClient(config as MinioClientOptions);
 }
 
-export function getTarObjectAsString({ bucket, key, client }: MinioRequestConfig) {
-  return new Promise<string>(async (resolve, reject) => {
-    try {
-      const stream = await getObjectStream({ bucket, key, client });
-      let contents = '';
-      // TODO: fix tar.Parse typing problem
-      stream.pipe(new (tar.Parse as any)()).on('entry', (entry: Stream) => {
-        entry.on('data', buffer => (contents += buffer.toString()));
-      });
-      stream.on('end', () => {
-        resolve(contents);
-      });
-    } catch (err) {
-      reject(err);
-    }
+/**
+ * Checks the magic number of a buffer to see if the mime type is a uncompressed
+ * tarball. The buffer must be of length 264 bytes or more.
+ * @param buf Buffer
+ */
+export function isTarball(buf: Buffer) {
+  if (!buf || buf.length < 264) {
+    return false;
+  }
+  const offset = 257;
+  const v1 = [0x75, 0x73, 0x74, 0x61, 0x72, 0x00, 0x30, 0x30];
+  const v0 = [0x75, 0x73, 0x74, 0x61, 0x72, 0x20, 0x20, 0x00];
+
+  return (
+    v1.reduce((res, curr, i) => res && curr === buf[offset + i], true) ||
+    v0.reduce((res, curr, i) => res && curr === buf[offset + i], true)
+  );
+}
+
+/**
+ * Returns a stream that extracts the first record of a tarball if the source
+ * stream is a tarball, otherwise just pipe the content as is.
+ */
+export function maybeTarball(): Transform {
+  return peek(
+    { newline: false, maxBuffer: 264 },
+    (data: Buffer, swap: (error?: Error, parser?: Transform) => void) => {
+      if (isTarball(data)) swap(null, extractFirstTarRecordAsStream());
+      else swap(null, new PassThrough());
```

Comment:
So we also support uncompressed artifacts after this change?
Great work!

Reply:
Yeah. The flow is
- if compressed, then deflate or gunzip, else just stream as normal
- if tarball, stream 1st record, else just stream everything